EIPS/eip-3074.md

```diff
@@ -27,125 +27,103 @@ While it is possible to emulate sponsored transactions (ex. [Gas Station Network
- - `retOffset: uint256`, `retLength: uint256` - region of memory filled with the return data from the call into the callee; and
- - `yParity: uint8`, `r: bytes32`, `s: bytes32` - signature for the package.
-
-#### Outputs
-
-`CALLFROM` pushes the following two values onto the stack:
+#### Behavior
 
-| Stack      | Value     |
-| ---------- | --------- |
-| `top - 0`  | `valid`   |
-| `top - 1`  | `success` |
+The arguments (`yParity`, `r`, `s`) are interpreted as an ECDSA signature of the form `secp256k1(keccak256(TYPE || abi.encode(invoker, commit)))`, where:
+ - `invoker` is the address of the contract executing `AUTHORIZE`.
+ - `commit` is one of the arguments passed into `AUTHORIZE` and is a 32-byte value that can be used to commit to specific additional validity condition in the invoker's pre-processing logic.
+
+If the signature is valid, the `signerAddress` is recovered. If `signerAddress != tx.origin`, the context variable `authorizedAccount` is set to `signerAddress`.
+In any other case, i.e. if the signature is invalid or `signerAddress == tx.origin`, `authorizedAccount` is reset to the zero address.
+
+`AUTHORIZE` returns the stack element `authorizedAccount`, which is set to the context variable of the same name.
 
+#### Gas Cost
 
-##### Output: `valid`
+The gas cost for `AUTHORIZE` is `3000`, which is the same cost as for the existing ecrecover precompile.
 
-`valid` must be zero in the following cases:
- - Invalid signature
- - The address recovered from `yParity`, `r`, and `s` does not match `sponsee`
- - The balance of the invoker is less than `value`
+### `AUTHORIZEDCALL` (`0xf9`)
 
-`valid` must be a one in all other cases.
+A new opcode `AUTHORIZEDCALL` is created at `0xf9`. It takes in 7 stack elements and returns 1 stack element. It closely mirrors the behavior of the existing `CALL` (`0xF1`) instruction.
 
-##### Output: `success`
+#### Input
+
+| Stack     | Value        |
+| --------- | ------------ |
+| `top - 0` | `gas`        |
+| `top - 1` | `callee`     |
+| `top - 2` | `value`      |
+| `top - 3` | `argsOffset` |
+| `top - 4` | `argsLength` |
+| `top - 5` | `retOffset`  |
+| `top - 6` | `retLength`  |
+
+#### Output
+
+| Stack      | Value     |
+| ---------- | --------- |
+| `top - 0`  | `success` |
 
-`success` must be zero in the following cases:
- - `success == 0`
- - the code execution failed due to an exceptional halting or revert
- - call depth limit has been reached
+#### Behavior
 
-`success` must be a one in all other cases.
+`AUTHORIZEDCALL` has the same execution logic as `CALL`, except for:
 
-##### Output: Return Data
+ - If `authorizedAccount` is the zero address, no call is made and `success` is set to `0`. Otherwise, the caller address for the call is set to `authorizedAccount`.
+ - The call value is deducted from the `authorizedAccount` balance. If that balance is insufficient, no call is made and `success` is set to `0`.
```

Comment:
I don't think this should be in the EIP. It is opening up a can of worms that I don't think we should open right now. Having the invoker maintain a balance will be good enough for 90% of cases.